libqtile/widget/redshift.py

```diff
@@ -0,0 +1,476 @@
+        else:
+            text = self._lines[self._line_index]
+        self.update(text)
+
+    def _format_temp_text(self) -> str:
+        """
+        Format the temperature text.
+        """
+        return self.temperature_fmt.format(temperature=self.temperature)
+
+    def _format_brightness_text(self) -> str:
+        """
+        Format the brightness text.
+        """
+        return self.brightness_fmt.format(brightness=self.brightness)
+
+    def _format_gamma_text(self) -> str:
+        """
+        Format the gamma text.
+        """
+        return self.gamma_val._format_gamma(self.gamma_fmt)
+
+    def _format_first_text(self, txt: str) -> str:
+        """
+        Format the enabled/disabled text (aka first_text).
+        """
+        return txt.format(
+            brightness=self.brightness,
+            temperature=self.temperature,
+            gamma_red=self.gamma_red,
+            gamma_green=self.gamma_green,
+            gamma_blue=self.gamma_blue,
+            is_enabled=self.is_enabled,
+        )
+
+    def _assert_brightness(self):
+        """
+        assert that self.brightness is within
+        the correct limits
+        """
+        assert (
+            self.brightness >= self.brightness_lower_lim
+            and self.brightness <= self.brightness_upper_lim
+        ), f"redshift: self.brightness is not initialised within the acceptable range, see the widget defaults docs: {self.brightness}"
+
+    def _assert_gamma(self) -> GammaGroup:
+        """
+        assert that self.gamma is within
+        the correct limits. If it is, produce
+        a GammaGroup instance
+        """
+        gamma_vals = [
+            self.gamma_red,
+            self.gamma_green,
+            self.gamma_blue,
+        ]
+        gamma_group = GammaGroup(self.gamma_red, self.gamma_green, self.gamma_blue)
+        for _, val in enumerate(gamma_vals):
+            assert (
+                val >= self.gamma_lower_lim and val <= self.gamma_upper_lim
+            ), f"redshift: self.gamma_red, self.gamma_green or self.gamma_blue have not been initialised within the acceptable range, see the docs: {gamma_group}"
+
+        return gamma_group
+
+    def _assert_temperature(self):
+        """
+        assert that self.temperature is within
+        the correct limits
+        """
+        assert (
+            self.temperature >= self.temperature_lower_lim
+            and self.temperature <= self.temperature_upper_lim
+        ), f"redshift: self.temperature is not initialised within the acceptable range, see the docs: {self.temperature}"
+
+    @expose_command
+    def reset_redshift(self):
+        """
+        Call reset on redshift to reset to default settings.
+        """
+        try:
+            subprocess.run(["redshift", "-x"], check=True)
+        except subprocess.CalledProcessError as e:
+            raise RuntimeError(f"redshift: could not enable redshift: {e}")
```

Comment:
We should log the error and display an error message rather than raising an error.